in_node_exporter_metrics: add support for thermal_zone.

pwhelan opened the pull request:
# Summary

This patch adds support for reading temperature values from `/sys/calss/thermal_zone`. The labels are replicated from the same behaviour as the prometheus node_exporter.

This plugin provides thermal reporting on linux/arm64, especially for raspberry pi 4 and other similar SBCs.

These sensors are already implemented in the prometheus node_exporter_metrics: https://github.com/prometheus/node_exporter/blob/ed1b8e3d88851806627e4f8262ee26232ca56c2c/collector/thermal_zone_linux.go#L31.

----
Enter `[N/A]` in the box, if an item is not applicable to your change.

**Testing**
Before we can approve your change; please submit the following in a comment:
- [x] Example configuration file for the change
- [x] Debug log output from testing the change
<!--  
Please refer to the Developer Guide for instructions on building Fluent Bit with Valgrind support: 
https://github.com/fluent/fluent-bit/blob/master/DEVELOPER_GUIDE.md#valgrind
Invoke Fluent Bit and Valgrind as: $ valgrind --leak-check=full ./bin/fluent-bit <args>
-->
- [x] Attached [Valgrind](https://valgrind.org/docs/manual/quick-start.html) output that shows no leaks or memory corruption was found

If this is a change to packaging of containers or native binaries then please confirm it works for all targets.
- [ ] Run [local packaging test](./packaging/local-build-all.sh) showing all targets (including any new ones) build.
- [ ] Set `ok-package-test` label to test for all targets (requires maintainer to do).

**Documentation**
<!-- Docs can be edited at https://github.com/fluent/fluent-bit-docs -->
- [ ] Documentation required for this feature

<!--  Doc PR (not required but highly recommended) -->

**Backporting**
<!--
PRs targeting the default master branch will go into the next major release usually.
If this PR should be backported to the current or earlier releases then please submit a PR for that particular branch.
-->
- [ ] Backport to latest stable release.

<!--  Other release PR (not required but highly recommended for quick turnaround) -->
----

Fluent Bit is licensed under Apache 2.0, by submitting this pull request I understand that this code will be released under the terms of that license.

pwhelan commented:
Here is a run using valgrind:

```bash
valgrind ./bin/fluent-bit -i node_exporter_metrics -p me
trics=thermal_zone -o stdout -m '*' -o exit -m '*' -f 1
==515970== Memcheck, a memory error detector
==515970== Copyright (C) 2002-2022, and GNU GPL'd, by Julian Seward et al.
==515970== Using Valgrind-3.21.0 and LibVEX; rerun with -h for copyright info
==515970== Command: ./bin/fluent-bit -i node_exporter_metrics -p metrics=thermal_zone -o stdout -m * -o exit -m * -f 1
==515970== 
Fluent Bit v2.1.5
* Copyright (C) 2015-2022 The Fluent Bit Authors
* Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
* https://fluentbit.io

[2023/06/06 19:46:51] [ info] [fluent bit] version=2.1.5, commit=3fdd42c6f2, pid=515970
[2023/06/06 19:46:51] [ info] [storage] ver=1.4.0, type=memory, sync=normal, checksum=off, max_chunks_up=128
[2023/06/06 19:46:51] [ info] [input:node_exporter_metrics:node_exporter_metrics.0] path.procfs = /proc
[2023/06/06 19:46:51] [ info] [cmetrics] version=0.6.1
[2023/06/06 19:46:51] [ info] [input:node_exporter_metrics:node_exporter_metrics.0] path.sysfs  = /sys
[2023/06/06 19:46:51] [ info] [ctraces ] version=0.3.1
[2023/06/06 19:46:51] [ info] [output:stdout:stdout.0] worker #0 started
[2023/06/06 19:46:51] [ info] [input:node_exporter_metrics:node_exporter_metrics.0] initializing
[2023/06/06 19:46:51] [ info] [input:node_exporter_metrics:node_exporter_metrics.0] storage_strategy='memory' (memory only)
[2023/06/06 19:46:51] [ info] [input:node_exporter_metrics:node_exporter_metrics.0] thread instance initialized
[2023/06/06 19:46:51] [ info] [sp] stream processor started
[2023/06/06 19:46:56] [error] [/home/pwhelan/Projects/personal/fluent-bit/plugins/in_node_exporter_metrics/ne_utils.c:117 errno=61] No data available
2023-06-06T23:46:56.265368405Z node_thermal_zone_temp{zone="0",type="acpitz"} = 16.800000000000001
2023-06-06T23:46:56.265368405Z node_thermal_zone_temp{zone="1",type="acpitz"} = 16.800000000000001
2023-06-06T23:46:56.265368405Z node_thermal_zone_temp{zone="2",type="acpitz"} = 16.800000000000001
2023-06-06T23:46:56.306184135Z node_cooling_device_cur_state{name="0",type="Processor"} = 0
2023-06-06T23:46:56.306184135Z node_cooling_device_cur_state{name="1",type="Processor"} = 0
2023-06-06T23:46:56.306184135Z node_cooling_device_cur_state{name="10",type="Processor"} = 0
2023-06-06T23:46:56.306184135Z node_cooling_device_cur_state{name="11",type="Processor"} = 0
2023-06-06T23:46:56.306184135Z node_cooling_device_cur_state{name="12",type="Processor"} = 0
2023-06-06T23:46:56.306184135Z node_cooling_device_cur_state{name="13",type="Processor"} = 0
2023-06-06T23:46:56.306184135Z node_cooling_device_cur_state{name="14",type="Processor"} = 0
2023-06-06T23:46:56.306184135Z node_cooling_device_cur_state{name="15",type="Processor"} = 0
2023-06-06T23:46:56.306184135Z node_cooling_device_cur_state{name="16",type="Processor"} = 0
2023-06-06T23:46:56.306184135Z node_cooling_device_cur_state{name="17",type="Processor"} = 0
2023-06-06T23:46:56.306184135Z node_cooling_device_cur_state{name="18",type="Processor"} = 0
2023-06-06T23:46:56.306184135Z node_cooling_device_cur_state{name="19",type="Processor"} = 0
2023-06-06T23:46:56.306184135Z node_cooling_device_cur_state{name="2",type="Processor"} = 0
2023-06-06T23:46:56.306184135Z node_cooling_device_cur_state{name="20",type="Processor"} = 0
2023-06-06T23:46:56.306184135Z node_cooling_device_cur_state{name="21",type="Processor"} = 0
2023-06-06T23:46:56.306184135Z node_cooling_device_cur_state{name="22",type="Processor"} = 0
2023-06-06T23:46:56.306184135Z node_cooling_device_cur_state{name="23",type="Processor"} = 0
2023-06-06T23:46:56.306184135Z node_cooling_device_cur_state{name="3",type="Processor"} = 0
2023-06-06T23:46:56.306184135Z node_cooling_device_cur_state{name="4",type="Processor"} = 0
2023-06-06T23:46:56.306184135Z node_cooling_device_cur_state{name="5",type="Processor"} = 0
2023-06-06T23:46:56.306184135Z node_cooling_device_cur_state{name="6",type="Processor"} = 0
2023-06-06T23:46:56.306184135Z node_cooling_device_cur_state{name="7",type="Processor"} = 0
2023-06-06T23:46:56.306184135Z node_cooling_device_cur_state{name="8",type="Processor"} = 0
2023-06-06T23:46:56.306184135Z node_cooling_device_cur_state{name="9",type="Processor"} = 0
2023-06-06T23:46:56.306184135Z node_cooling_device_max_state{name="0",type="Processor"} = 10
2023-06-06T23:46:56.306184135Z node_cooling_device_max_state{name="1",type="Processor"} = 10
2023-06-06T23:46:56.306184135Z node_cooling_device_max_state{name="10",type="Processor"} = 10
2023-06-06T23:46:56.306184135Z node_cooling_device_max_state{name="11",type="Processor"} = 10
2023-06-06T23:46:56.306184135Z node_cooling_device_max_state{name="12",type="Processor"} = 10
2023-06-06T23:46:56.306184135Z node_cooling_device_max_state{name="13",type="Processor"} = 10
2023-06-06T23:46:56.306184135Z node_cooling_device_max_state{name="14",type="Processor"} = 10
2023-06-06T23:46:56.306184135Z node_cooling_device_max_state{name="15",type="Processor"} = 10
2023-06-06T23:46:56.306184135Z node_cooling_device_max_state{name="16",type="Processor"} = 10
2023-06-06T23:46:56.306184135Z node_cooling_device_max_state{name="17",type="Processor"} = 10
2023-06-06T23:46:56.306184135Z node_cooling_device_max_state{name="18",type="Processor"} = 10
2023-06-06T23:46:56.306184135Z node_cooling_device_max_state{name="19",type="Processor"} = 10
2023-06-06T23:46:56.306184135Z node_cooling_device_max_state{name="2",type="Processor"} = 10
2023-06-06T23:46:56.306184135Z node_cooling_device_max_state{name="20",type="Processor"} = 10
2023-06-06T23:46:56.306184135Z node_cooling_device_max_state{name="21",type="Processor"} = 10
2023-06-06T23:46:56.306184135Z node_cooling_device_max_state{name="22",type="Processor"} = 10
2023-06-06T23:46:56.306184135Z node_cooling_device_max_state{name="23",type="Processor"} = 10
2023-06-06T23:46:56.306184135Z node_cooling_device_max_state{name="3",type="Processor"} = 10
2023-06-06T23:46:56.306184135Z node_cooling_device_max_state{name="4",type="Processor"} = 10
2023-06-06T23:46:56.306184135Z node_cooling_device_max_state{name="5",type="Processor"} = 10
2023-06-06T23:46:56.306184135Z node_cooling_device_max_state{name="6",type="Processor"} = 10
2023-06-06T23:46:56.306184135Z node_cooling_device_max_state{name="7",type="Processor"} = 10
2023-06-06T23:46:56.306184135Z node_cooling_device_max_state{name="8",type="Processor"} = 10
2023-06-06T23:46:56.306184135Z node_cooling_device_max_state{name="9",type="Processor"} = 10
^C[2023/06/06 19:46:59] [engine] caught signal (SIGINT)
[2023/06/06 19:46:59] [ warn] [engine] service will shutdown in max 5 seconds
[2023/06/06 19:47:00] [ info] [engine] service has stopped (0 pending tasks)
[2023/06/06 19:47:00] [ warn] [input:node_exporter_metrics:node_exporter_metrics.0] Unknown metrics: thermal_zone
[2023/06/06 19:47:00] [ info] [output:stdout:stdout.0] thread worker #0 stopping...
[2023/06/06 19:47:00] [ info] [output:stdout:stdout.0] thread worker #0 stopped
==515970== 
==515970== HEAP SUMMARY:
==515970==     in use at exit: 0 bytes in 0 blocks
==515970==   total heap usage: 3,802 allocs, 3,802 frees, 2,640,178 bytes allocated
==515970== 
==515970== All heap blocks were freed -- no leaks are possible
==515970== 
==515970== For lists of detected and suppressed errors, rerun with: -s
==515970== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
```

There are no memory leaks as far as I can ascertain.

cosmo0920 commented on plugins/in_node_exporter_metrics/ne_thermalzone_linux.c:
```diff
@@ -0,0 +1,242 @@
+
+
+/*
+ * See kernel documentation for a description:
+ * https://www.kernel.org/doc/html/latest/filesystems/proc.html
+ *
+ * user: normal processes executing in user mode
+ * nice: niced processes executing in user mode
+ * system: processes executing in kernel mode
+ * idle: twiddling thumbs
+ * iowait: In a word, iowait stands for waiting for I/O to complete. But there are several problems:
+ * irq: servicing interrupts
+ * softirq: servicing softirqs
+ * steal: involuntary wait
+ * guest: running a normal guest
+ * guest_nice: running a niced guest
+ *
+ * Ensure to pick the correct version of the documentation, older versions here:
+ * https://github.com/torvalds/linux/tree/master/Documentation
```
BTW, these lines should be updated to reflect for thermal_zone specifications?

leonardo-albertovich commented on plugins/in_node_exporter_metrics/ne_thermalzone_linux.c:
```diff
@@ -0,0 +1,242 @@
+            num = entry->str;
+        }
+
+        cmt_gauge_set(ctx->thermalzone_temp, ts, ((double)temp)/1000.0,
```
Add the missing space after the type cast and use appropriate names for the variables.

pwhelan replied:
the variable is the temp(erature), seems appropriate to me.

leonardo-albertovich commented on plugins/in_node_exporter_metrics/ne_thermalzone_linux.c:
```diff
@@ -0,0 +1,242 @@
+    struct flb_slist_entry *entry;
+    flb_sds_t type;
+    char *num;
+    flb_sds_t syspfx = flb_sds_create_size(strlen(COOLING_DEVICE_PATTERN) + 
```
Same as in line 97

leonardo-albertovich commented on plugins/in_node_exporter_metrics/ne_thermalzone_linux.c:
```diff
@@ -0,0 +1,242 @@
+    char *num;
+    flb_sds_t syspfx = flb_sds_create_size(strlen(COOLING_DEVICE_PATTERN) + 
+                                           strlen(ctx->path_sysfs) + 8);
+    const char *pattern = "/class/thermal/cooling_device[0-9]*";
```
Define and use a constant instead of this.

leonardo-albertovich commented on plugins/in_node_exporter_metrics/ne_thermalzone_linux.c:
```diff
@@ -0,0 +1,242 @@
+    struct mk_list list;
+    struct flb_slist_entry *entry;
+    flb_sds_t type;
+    flb_sds_t syspfx = flb_sds_create_size(strlen(THERMAL_ZONE_PATTERN) + 
```
When initializing `syspfx` check the result of `flb_sds_create_size`

leonardo-albertovich commented on plugins/in_node_exporter_metrics/ne_thermalzone_linux.h:
```diff
@@ -0,0 +1,29 @@
+int ne_thermalzone_update_thermal_zones(struct flb_ne *ctx);
+int ne_thermalzone_update_cooling_devices(struct flb_ne *ctx);
+
+#endif // FLB_IN_NE_THERMALZONE_LINUX_H
```
Remove this comment

leonardo-albertovich commented on plugins/in_node_exporter_metrics/ne_utils.c:
```diff
@@ -178,6 +178,79 @@ int ne_utils_file_read_lines(const char *mount, const char *path, struct mk_list
+ * Read a file and every non-empty line is stored as a flb_slist_entry in the
+ * given list.
+ */
+int ne_utils_file_read_sds(const char *mount, 
```
Rename the variables with meaningful names.

leonardo-albertovich replied:
Add the missing safety checks for  `flb_sds_cat_safe`

pwhelan replied:
The names of the arguments follow the convention used in the other functions in the same file.

pwhelan commented on plugins/in_node_exporter_metrics/ne_thermalzone_linux.c:
```diff
@@ -0,0 +1,242 @@
+
+int ne_thermalzone_init(struct flb_ne *ctx)
+{
+    struct cmt_gauge *g;
```
Doing so will make the lines way too long to make it possible to group the parameters in any legible way if wrapping up to 80 columns.

edsiper commented:
@pwhelan @leonardo-albertovich is this ready to go ?

pwhelan commented:
> @pwhelan @leonardo-albertovich is this ready to go ?

I just added the checks for the calls to flb_sds_cat_safe. I kept the variable names the same since it follows the conventions used throughout the `ne_utils.c` file. I don't understand myself really what the original logic was for `join_a` and `join_b` so changing the names to something more sensical is beyond me. @cosmo0920 might have some insight there.

@edsiper As far as I'm concerned it's ready to go.

cosmo0920 reviewed (commented):
I'm not having much insight of join_a and join_b logic but I did the same implementation for https://github.com/fluent/fluent-bit/pull/7522/files#diff-f32396faf19457b78c990f276acfbb6dde174a2ba49f99197ce827d312659641R212-R216.
This could prevent duplicating a sysfs mount point.
I checked the PR for Intel and AMD(Ryzen) platforms and it works well.

pwhelan commented:
@leonardo-albertovich I'd be happy to open a new PR that renames `join_a` and `join_b` to something more understandable but I think it falls out of the scope of this PR.

leonardo-albertovich commented:
Could you please go through the changes I requested, double check them and mark them as resolved if they are or make a comment if there are any issues?

pwhelan commented:
> Could you please go through the changes I requested, double check them and mark them as resolved if they are or make a comment if there are any issues?

I have double checked all of the requested changes. In doing so I spotted some code which was duplicated and removed it. I also renamed the `ts` variable, which we use for timestamps throughout the entire code base and renamed them to `tstamp`.

cosmo0920 commented on plugins/in_node_exporter_metrics/ne.c:
```diff
@@ -1009,9 +1057,15 @@ static struct flb_config_map config_map[] = {
     },
 
+    {
+     FLB_CONFIG_MAP_TIME, "collector.thermalzone.scrape_interval", "0",
```
This is just for my curiosity. 
Should we align the name of this thermal_zone related parameter to the actual metrics name?

pwhelan replied:
Probably for the best. I'll change it.

cosmo0920 reviewed (commented):
Looks good.
I added a nitpick comment.

Also, we need to sort the conflict:

```diff
<<<<<<< node-exporter-thermal_zone
    ctx->coll_thermalzone_fd = -1;
=======
    ctx->coll_systemd_fd = -1;
>>>>>>> master
```

cosmo0920 commented on plugins/in_node_exporter_metrics/ne.c:
```diff
@@ -740,6 +763,26 @@ static int in_ne_init(struct flb_input_instance *in,
+                else if (strncmp(entry->str, "thermal_zone", 12) == 0) {
+                    if (ctx->thermalzone_scrape_interval == 0) {
+                        flb_plg_debug(ctx->ins, "enabled metrics %s", entry->str);
+                        metric_idx = 14;
```
Could you increment the index value to 15? This is because my patch's metrics is injected as index of 14.

cosmo0920 commented on plugins/in_node_exporter_metrics/ne_thermalzone_linux.h:
```diff
@@ -0,0 +1,34 @@
+#ifndef FLB_IN_NE_THERMALZONE_LINUX_H
+#define FLB_IN_NE_THERMALZONE_LINUX_H
+
+#include "ne.h"
+
+#define THERMAL_ZONE_BASE "/class/thermal/thermal_zone"
+#define THERMAL_ZONE_PATTERN THERMAL_ZONE_BASE "[0-9]*"
+#define COOLING_DEVICE_BASE "/class/thermal/cooling_device"
+#define COOLING_DEVICE_PATTERN COOLING_DEVICE_BASE "[0-9]*"
+
+int ne_thermalzone_init(struct flb_ne *ctx);
+int ne_thermalzone_update_thermal_zones(struct flb_ne *ctx);
+int ne_thermalzone_update_cooling_devices(struct flb_ne *ctx);
```
Shouldn't we take care of macOS(darwin)? We already has merged https://github.com/fluent/fluent-bit/pull/7947 PR which supports macOS. So, we have to disable this metrics on macOS.

pwhelan replied:
The same needs to be done for at least:
- ne_stat_linux.c
- ne_vmstat_linux.c
- ne_filefd_linux.c

pwhelan replied:
I copied ne_filesystem.[c,h] to make the thermal_zone implementation cross platform capable and leave room for other implementations on other OSes.

cosmo0920 reviewed (changes requested):
I found a few issues on rebased patches. Could you take care of them?

cosmo0920 reviewed (approved):
Looks good to me.

cosmo0920 commented:
This could be adopted for more pluggable structure. @edsiper Any missing pieces of getting to be merged?

cosmo0920 commented:
@pwhelan Any chance to register the corresponding document for this PR?

pwhelan commented:
> This could be adopted for more pluggable structure. @edsiper Any missing pieces of getting to be merged?

I already adapted to @nokute78's new pluggable structure.

edsiper commented:
thanks everyone, merging this now.

note: I will squash the commits since all belongs to the new thermal_zone collector functionality
